File: src/common/api/worker/facades/UserFacade.ts
import { GroupType } from "../../common/TutanotaConstants"
import { Aes256Key, AesKey, decryptKey } from "@tutao/tutanota-crypto"
import { assertNotNull, KeyVersion } from "@tutao/tutanota-utils"
import { ProgrammingError } from "../../common/error/ProgrammingError"
import { createWebsocketLeaderStatus, GroupMembership, User, UserGroupKeyDistribution, WebsocketLeaderStatus } from "../../entities/sys/TypeRefs"
import { LoginIncompleteError } from "../../common/error/LoginIncompleteError"
import { isSameId } from "../../common/utils/EntityUtils.js"
import { KeyCache } from "./KeyCache.js"
import { CryptoWrapper, VersionedKey } from "../crypto/CryptoWrapper.js"
import { CryptoError } from "@tutao/tutanota-crypto/error.js"
import { checkKeyVersionConstraints, parseKeyVersion } from "./KeyLoaderFacade.js"

export interface AuthDataProvider {
	/**
	 * @return The map which contains authentication data for the logged-in user.
	 */
	createAuthHeaders(): Dict

	isFullyLoggedIn(): boolean
}

/** Holder for the user and session-related data on the worker side. */
export class UserFacade implements AuthDataProvider {
	private user: User | null = null
	private accessToken: string | null = null
	private leaderStatus!: WebsocketLeaderStatus

	constructor(
		private readonly keyCache: KeyCache,
		private readonly cryptoWrapper: CryptoWrapper,
	) {
		this.reset()
	}

	// Login process is somehow multi-step, and we don't use a separate network stack for it. So we have to break up setters.
	// 1. We need to download user. For that we need to set access token already (to authenticate the request for the server as it is passed in headers).
	// 2. We need to get group keys. For that we need to unlock userGroupKey with userPassphraseKey
	// so this leads to this steps in UserFacade:
	// 1. Access token is set
	// 2. User is set
	// 3. UserGroupKey is unlocked
	setAccessToken(accessToken: string | null) {
		this.accessToken = accessToken
	}

	getAccessToken(): string | null {
		return this.accessToken
	}

	setUser(user: User) {
		if (this.accessToken == null) {
			throw new ProgrammingError("invalid state: no access token")
		}
		this.user = user
	}

	unlockUserGroupKey(userPassphraseKey: AesKey) {
		if (this.user == null) {
			throw new ProgrammingError("Invalid state: no user")
		}
		const userGroupMembership = this.user.userGroup
		const currentUserGroupKey = {
			version: parseKeyVersion(userGroupMembership.groupKeyVersion),
			object: decryptKey(userPassphraseKey, userGroupMembership.symEncGKey),
		}
		this.keyCache.setCurrentUserGroupKey(currentUserGroupKey)
		this.setUserDistKey(currentUserGroupKey.version, userPassphraseKey)
	}

	setUserDistKey(currentUserGroupKeyVersion: KeyVersion, userPassphraseKey: AesKey) {
		if (this.user == null) {
			throw new ProgrammingError("Invalid state: no user")
		}
		// Why this magic + 1? Because we don't have access to the new version number when calling this function so we compute it from the current one
		const newUserGroupKeyVersion = checkKeyVersionConstraints(currentUserGroupKeyVersion + 1)
		const userGroupMembership = this.user.userGroup
		const legacyUserDistKey = this.deriveLegacyUserDistKey(userGroupMembership.group, userPassphraseKey)
		const userDistKey = this.deriveUserDistKey(userGroupMembership.group, newUserGroupKeyVersion, userPassphraseKey)
		this.keyCache.setLegacyUserDistKey(legacyUserDistKey)
		this.keyCache.setUserDistKey(userDistKey)
	}

	/**
	 * Derives a distribution key from the password key to share the new user group key of the user to their other clients (apps, web etc)
	 * This is a fallback function that gets called when the output key of `deriveUserDistKey` fails to decrypt the new user group key
	 * @deprecated
	 * @param userGroupId user group id of the logged in user
	 * @param userPasswordKey current password key of the user
	 */
	deriveLegacyUserDistKey(userGroupId: Id, userPasswordKey: AesKey): AesKey {
		// we prepare a key to encrypt potential user group key rotations with
		// when passwords are changed clients are logged-out of other sessions
		// this key is only needed by the logged-in clients, so it should be reliable enough to assume that userPassphraseKey is in sync

		// we bind this to userGroupId and the domain separator userGroupKeyDistributionKey from crypto package
		// the hkdf salt does not have to be secret but should be unique per user and carry some additional entropy which sha256 ensures

		return this.cryptoWrapper.deriveKeyWithHkdf({
			salt: userGroupId,
			key: userPasswordKey,
			context: "userGroupKeyDistributionKey",
		})
	}

	/**
	 * Derives a distribution to share the new user group key of the user to their other clients (apps, web etc)
	 * @param userGroupId user group id of the logged in user
	 * @param newUserGroupKeyVersion the new user group key version
	 * @param userPasswordKey current password key of the user
	 */
	deriveUserDistKey(userGroupId: Id, newUserGroupKeyVersion: KeyVersion, userPasswordKey: AesKey): Aes256Key {
		return this.cryptoWrapper.deriveKeyWithHkdf({
			salt: `userGroup: ${userGroupId}, newUserGroupKeyVersion: ${newUserGroupKeyVersion}`,
			key: userPasswordKey,
			// Formerly,this was not bound to the user group key version.
			context: "versionedUserGroupKeyDistributionKey",
		})
	}

	async updateUser(user: User) {
		if (this.user == null) {
			throw new ProgrammingError("Update user is called without logging in. This function is not for you.")
		}
		this.user = user
		await this.keyCache.removeOutdatedGroupKeys(user)
	}

	getUser(): User | null {
		return this.user
	}

	/**
	 * @return The map which contains authentication data for the logged-in user.
	 */
	createAuthHeaders(): Dict {
		return this.accessToken
			? {
					accessToken: this.accessToken,
				}
			: {}
	}

	getUserGroupId(): Id {
		return this.getLoggedInUser().userGroup.group
	}

	getAllGroupIds(): Id[] {
		let groups = this.getLoggedInUser().memberships.map((membership) => membership.group)
		groups.push(this.getLoggedInUser().userGroup.group)
		return groups
	}

	getCurrentUserGroupKey(): VersionedKey {
		// the userGroupKey is always written after the login to this.currentUserGroupKey
		//if the user has only logged in offline this has not happened
		const currentUserGroupKey = this.keyCache.getCurrentUserGroupKey()
		if (currentUserGroupKey == null) {
			if (this.isPartiallyLoggedIn()) {
				throw new LoginIncompleteError("userGroupKey not available")
			} else {
				throw new ProgrammingError("Invalid state: userGroupKey is not available")
			}
		}
		return currentUserGroupKey
	}

	getMembership(groupId: Id): GroupMembership {
		let membership = this.getLoggedInUser().memberships.find((g: GroupMembership) => isSameId(g.group, groupId))

		if (!membership) {
			throw new Error(`No group with groupId ${groupId} found!`)
		}

		return membership
	}

	hasGroup(groupId: Id): boolean {
		if (!this.user) {
			return false
		} else {
			return groupId === this.user.userGroup.group || this.user.memberships.some((m) => m.group === groupId)
		}
	}

	getGroupId(groupType: GroupType): Id {
		if (groupType === GroupType.User) {
			return this.getUserGroupId()
		} else {
			let membership = this.getLoggedInUser().memberships.find((m) => m.groupType === groupType)

			if (!membership) {
				throw new Error("could not find groupType " + groupType + " for user " + this.getLoggedInUser()._id)
			}

			return membership.group
		}
	}

	getGroupIds(groupType: GroupType): Id[] {
		return this.getLoggedInUser()
			.memberships.filter((m) => m.groupType === groupType)
			.map((gm) => gm.group)
	}

	isPartiallyLoggedIn(): boolean {
		return this.user != null
	}

	isFullyLoggedIn(): boolean {
		// We have userGroupKey, and we can decrypt any other key - we are good to go
		return this.keyCache.getCurrentUserGroupKey() != null
	}

	getLoggedInUser(): User {
		return assertNotNull(this.user)
	}

	setLeaderStatus(status: WebsocketLeaderStatus) {
		this.leaderStatus = status
		console.log("New leader status set:", status.leaderStatus)
	}

	isLeader(): boolean {
		return this.leaderStatus.leaderStatus
	}

	reset() {
		this.user = null
		this.accessToken = null
		this.keyCache.reset()
		this.leaderStatus = createWebsocketLeaderStatus({
			leaderStatus: false,
			// a valid applicationVersionSum and applicationTypesHash can only be provided by the server
			applicationVersionSum: null,
			applicationTypesHash: null,
		})
	}

	updateUserGroupKey(userGroupKeyDistribution: UserGroupKeyDistribution) {
		const userDistKey = this.keyCache.getUserDistKey()
		if (userDistKey == null) {
			console.log("could not update userGroupKey because distribution key is not available")
			return
		}
		let newUserGroupKeyBytes
		try {
			try {
				newUserGroupKeyBytes = decryptKey(userDistKey, userGroupKeyDistribution.distributionEncUserGroupKey)
			} catch (e) {
				if (e instanceof CryptoError) {
					// this might be due to old encryption with the legacy derivation of the distribution key
					// try with the legacy one instead
					const legacyUserDistKey = this.keyCache.getLegacyUserDistKey()
					if (legacyUserDistKey == null) {
						console.log("could not update userGroupKey because old legacy distribution key is not available")
						return
					}
					newUserGroupKeyBytes = decryptKey(legacyUserDistKey, userGroupKeyDistribution.distributionEncUserGroupKey)
				} else {
					throw e
				}
			}
		} catch (e) {
			// this may happen during offline storage synchronisation when the event queue contains user group key rotation and a password change.
			// We can ignore this error as we already have the latest user group key after connecting the offline client
			console.log(`Could not decrypt userGroupKeyUpdate`, e)
			return
		}
		const newUserGroupKey = {
			object: newUserGroupKeyBytes,
			version: parseKeyVersion(userGroupKeyDistribution.userGroupKeyVersion),
		}
		this.setNewUserGroupKey(newUserGroupKey)
	}

	/**
	 * Update the KeyCache with the newest user group key.
	 * NOTE: should only be used with a freshly generated key. For keys received from the server, use `updateUserGroupKey`
	 * @param userGroupKey
	 */
	public setNewUserGroupKey(userGroupKey: VersionedKey) {
		console.log(`updating userGroupKey. new version: ${userGroupKey.version}`)
		this.keyCache.setCurrentUserGroupKey(userGroupKey)
	}
}

File: tuta-sdk/rust/sdk/src/user_facade.rs
use crate::crypto::hkdf;
use crate::crypto::key::GenericAesKey;
use crate::crypto::key::VersionedAesKey;
use crate::crypto::sha256;
use crate::crypto::{Aes256Key, AES_256_KEY_SIZE};
use crate::entities::generated::sys::{GroupMembership, User};
use crate::groups::GroupType;
#[cfg_attr(test, mockall_double::double)]
use crate::key_cache::KeyCache;
use crate::util::{convert_version_to_u64, Versioned};
use crate::ApiCallError;
use crate::GeneratedId;
use base64::prelude::BASE64_STANDARD;
use base64::Engine;
use std::borrow::ToOwned;
use std::sync::{Arc, RwLock};

#[allow(unused)]
const USER_GROUP_KEY_DISTRIBUTION_KEY_INFO: &[u8] = b"userGroupKeyDistributionKey";

pub struct UserFacade {
	user: RwLock<Arc<User>>,
	#[allow(unused)]
	key_cache: Arc<KeyCache>,
}

/// UserFacade is tied to a logged in user.
#[cfg_attr(test, mockall::automock)]
#[allow(unused)]
impl UserFacade {
	#[allow(unused)]
	pub fn new(key_cache: Arc<KeyCache>, user: User) -> Self {
		UserFacade {
			user: RwLock::new(Arc::new(user)),
			key_cache,
		}
	}

	pub fn set_user(&self, user: User) {
		*self.user.write().unwrap() = Arc::new(user);
	}

	pub fn unlock_user_group_key(
		&self,
		user_passphrase_key: GenericAesKey,
	) -> Result<(), ApiCallError> {
		let user = self.get_user();
		let user_group_membership = &user.userGroup;
		let current_user_group_key = Versioned::new(
			user_passphrase_key
				.decrypt_aes_key(&user_group_membership.symEncGKey)
				.map_err(|e| ApiCallError::InternalSdkError {
					error_message: e.to_string(),
				})?,
			convert_version_to_u64(user_group_membership.groupKeyVersion),
		);
		self.key_cache
			.set_current_user_group_key(current_user_group_key);
		self.set_user_group_key_distribution_key(user_passphrase_key);
		Ok(())
	}

	pub fn key_cache(&self) -> Arc<KeyCache> {
		self.key_cache.clone()
	}

	fn set_user_group_key_distribution_key(&self, user_passphrase_key: GenericAesKey) {
		let user = self.get_user();
		let user_group_membership = &user.userGroup;
		let user_group_key_distribution_key = self.derive_user_group_key_distribution_key(
			&user_group_membership.group,
			user_passphrase_key,
		);
		self.key_cache
			.set_user_group_key_distribution_key(user_group_key_distribution_key)
	}

	// TODO: Add a test to check uint8ArrayToBase64 is correct;
	// there is a max length in the ts version, it seems to be a js thing, can we forego it here?
	fn derive_user_group_key_distribution_key(
		&self,
		user_group_id: &GeneratedId,
		user_passphrase_key: GenericAesKey,
	) -> Aes256Key {
		// we prepare a key to encrypt potential user group key rotations with
		// when passwords are changed clients are logged-out of other sessions
		// this key is only needed by the logged-in clients, so it should be reliable enough to assume that userPassphraseKey is in sync
		let user_group_id_hash = sha256(user_group_id.as_str().as_bytes());
		// we bind this to userGroupId and the domain separator USER_GROUP_KEY_DISTRIBUTION_KEY_INFO
		// the hkdf salt does not have to be secret but should be unique per user and carry some additional entropy which sha256 ensures
		let aes_key = hkdf(
			user_group_id_hash.as_slice(),
			BASE64_STANDARD
				.encode(user_passphrase_key.as_bytes())
				.as_bytes(),
			USER_GROUP_KEY_DISTRIBUTION_KEY_INFO,
			AES_256_KEY_SIZE,
		);

		Aes256Key::from_bytes(aes_key.as_slice()).expect("invalid derived key size")
	}

	pub async fn update_user(&self, user: User) {
		let user = Arc::new(user);
		*self.user.write().unwrap() = user.clone();
		self.key_cache
			.remove_outdated_group_keys(user.as_ref())
			.await;
	}

	pub fn get_user(&self) -> Arc<User> {
		self.user.read().unwrap().clone()
	}

	pub fn get_user_group_id(&self) -> GeneratedId {
		self.get_user().userGroup.group.clone()
	}

	#[allow(dead_code)] // Remove when implementing `generateUserAreaGroupData()`
	fn get_all_group_ids(&self) -> Vec<GeneratedId> {
		let mut groups: Vec<GeneratedId> = self
			.get_user()
			.memberships
			.iter()
			.map(|membership| membership.group.clone())
			.collect();
		groups.push(self.get_user().userGroup.group.clone());
		groups
	}

	pub fn get_current_user_group_key(&self) -> Option<VersionedAesKey> {
		self.key_cache.get_current_user_group_key()
	}

	pub fn get_membership_by_group_type(
		&self,
		group_type: GroupType,
	) -> Result<GroupMembership, ApiCallError> {
		let memberships = &self.get_user().memberships;
		let group_type = group_type as i64;
		memberships
			.iter()
			.find(|g| g.groupType == Some(group_type))
			.map(|m| m.to_owned())
			.ok_or_else(|| ApiCallError::InternalSdkError {
				error_message: format!("No group with groupType {} found!", group_type),
			})
	}

	pub fn has_group(&self, group_id: &GeneratedId) -> bool {
		&self.get_user().userGroup.group == group_id
			|| (self.get_user().memberships)
				.iter()
				.any(|m| group_id == &m.group)
	}

	#[allow(unused)]
	pub(crate) fn get_membership(
		&self,
		group_id: &GeneratedId,
	) -> Result<GroupMembership, ApiCallError> {
		let memberships = &self.get_user().memberships;
		memberships
			.iter()
			.find(|g| g.group == *group_id)
			.map(|m| m.to_owned())
			.ok_or_else(|| ApiCallError::InternalSdkError {
				error_message: format!("No group with groupId {} found!", group_id),
			})
	}
}

impl GroupMembership {
	#[must_use]
	pub fn group_type(&self) -> GroupType {
		match self.groupType {
			None => GroupType::Unknown,
			Some(group_type) => {
				GroupType::try_from(group_type as u64).unwrap_or(GroupType::Unknown)
			},
		}
	}
}
